metis-transformation/metis-transformation-service/src/main/java/eu/europeana/metis/transformation/service/XsltTransformer.java

```diff
@@ -1,84 +1,137 @@
+   * that will be discarded after the completion of this method. If multiple files are to be
+   * transformed, it is more efficient to manually create an instance of this class using one of the
+   * constructors and then call its {@link #transform(byte[])} method for each of those files.
    *
-   * @param xsltUrl url to xslt file
-   * @param fileContent content of the file
-   * @param datasetIdValue value that will be injected to datasetId field
-   * @return transformed file
+   * @param xsltUrl The URL of the XSLT file.
+   * @param fileContent The file to be transformed.
+   * @param datasetId The value that will be injected to the datasetId field in the XSLT. Can be
+   *        null.
+   * @return The transformed file.
    * @throws TransformationException if a problem occurred during transformation
    */
-  public StringWriter transform(String xsltUrl, byte[] fileContent, String datasetIdValue)
+  public static StringWriter transform(String xsltUrl, byte[] fileContent, String datasetId)
       throws TransformationException {
-    try (final InputStream xsltStream = new URL(xsltUrl).openStream()) {
-      final Transformer transformer = getTransformer(xsltUrl, xsltStream);
-      addParameter(transformer, "datasetId", datasetIdValue);
-      try (final InputStream contentStream = new ByteArrayInputStream(fileContent)) {
-        return executeTransformation(transformer, contentStream);
-      }
-    } catch (IOException | TransformerException e) {
-      LOGGER.error("Exception during transformation", e);
-      throw new TransformationException(e);
-    }
+    return new XsltTransformer(xsltUrl, datasetId).transform(fileContent);
   }
 
-  private Transformer getTransformer(String xsltUrl, InputStream xsltStream)
-      throws TransformerConfigurationException {
-    if (cache.containsKey(xsltUrl)) {
-      return cache.get(xsltUrl);
-    } else {
-      Source xslDoc = new StreamSource(xsltStream);
-      TransformerFactory transformerFactory = new TransformerFactoryImpl();
-      Transformer transformer = transformerFactory.newTransformer(xslDoc);
-      cache.put(xsltUrl, transformer);
-      return transformer;
+  private static Templates getTemplates(String xsltUrl)
+      throws TransformerConfigurationException, IOException {
+
+    // Check if the cache already contains this URL.
+    synchronized (cache) {
+      if (cache.containsKey(xsltUrl)) {
+        return cache.get(xsltUrl);
+      }
     }
-  }
 
-  private StringWriter executeTransformation(Transformer transformer, InputStream contentStream)
-      throws TransformerException {
-    Source xmlDoc = new StreamSource(contentStream);
-    StringWriter writer = new StringWriter();
-    transformer.transform(xmlDoc, new StreamResult(writer));
-    return writer;
-  }
+    // If it doesn't, resolve the URL and create the compiled transformation.
+    final TransformerFactory transformerFactory = new TransformerFactoryImpl();
+    final Templates templates;
```

Comment:
If the final variable is not set immediately then this usually is a code smell on SonarQube.